crates/core/c8y_api/src/smartrest/operations.rs

```diff
@@ -10,18 +29,326 @@ use std::collections::HashSet;
 use std::fs;
 use std::path::Path;
 use std::path::PathBuf;
+use std::sync::Arc;
 use std::time::Duration;
 use tedge_api::substitution::Record;
 use tedge_config::TopicPrefix;
+use tedge_utils::file;
 use tracing::error;
 use tracing::warn;
 
 use super::payload::SmartrestPayload;
+use super::topic::C8yTopic;
 
 const DEFAULT_GRACEFUL_TIMEOUT: Duration = Duration::from_secs(3600);
 const DEFAULT_FORCEFUL_TIMEOUT: Duration = Duration::from_secs(60);
 
+type ExternalId = String;
+type ExternalIdRef = str;
+
 type OperationName = String;
+type OperationNameRef = str;
+
+/// Used to hold and query supported operations for all devices.
+pub struct SupportedOperations {
+    /// External ID of the main device.
+    ///
+    /// Required because when given an external id of the main device when creating an operation, we want to create it
+    /// in a main directory, instead of a subdirectory.
```

Comment:
Okay. It makes sense for this PR focus on code refactoring without any user-visible changes, but this asymmetry between the main device and the child devices makes things more complicated than they should.

Reply:
I don't fully get what you mean. Do you mean this PR specifically, or the codebase more generally?

One maybe unnecessary change that I did was instead of having "operations for the main device" in one field and "operations for child devices" in a hashmap, i put it all in a single hashmap, so that `CumulocityConverter` only passes along the external id and doesn't have to care if a device is a child device or a main device. In retrospect the functions can still take an external id but the operations for the main device can be made a normal field instead of a hashmap entry, since "operations for the main device" is something that always exists and putting it in a hashmap where it can be removed introduces an extra failure mode.

Reply:
My point is in no case to disagree with this PR which is a real improvement reducing the accidental complexity of the c8y mapper.

My comment is coming from the fact that you (correctly) feel the need to add a comment on why we need the main device external id here (while we don't for the child devices).